### PR TITLE
Gmm

### PR DESCRIPTION
**Pull request by albertzhu01**

### CODE REVIEW

File: nequip/utils/gmm.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+@torch.jit.script
+def _compute_log_det_cholesky(
+    matrix_chol: torch.Tensor, covariance_type: str, n_features: int
+):
+    """Compute the log-det of the cholesky decomposition of matrices."""
+
+    assert covariance_type in ("full",)
+    n_components = matrix_chol.size(dim=0)
+    log_det_chol = torch.sum(
+        torch.log(matrix_chol.view(n_components, -1)[:, :: n_features + 1]), dim=1
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:17]: add link to sklearn code for this mysterious calculation?

File: nequip/utils/gmm.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    assert covariance_type in ("full",)
+    n_samples, n_features = X.size(dim=0), X.size(dim=1)
+    n_components = means.size(dim=0)
+
+    # det(precision_chol) = -0.5 * det(precision)
+    log_det = _compute_log_det_cholesky(precisions_chol, covariance_type, n_features)
+
+    # TODO: understand this
+    log_prob = torch.empty((n_samples, n_components))
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:40]: replace with vectorized version

File: nequip/utils/gmm.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    #     torch.square(
+    #         torch.matmul(X, precisions_chol.transpose(1, 2))
+    #         - torch.matmul(means, precisions_chol.transpose(1, 2))
+    #     ),
+    #     dim=1,
+    # )
+
+    # Since we are using the precision of the Cholesky decomposition,
+    # `-0.5 * log_det` becomes `+ log_det`
+    return -0.5 * (n_features * torch.log(torch.tensor(2 * np.pi)) + log_prob) + log_det
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:58]: use `math.pi`, and remove `torch.tensor`.

File: nequip/utils/gmm.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        covariance_type: str = "full",
+    ):
+        super(GaussianMixture, self).__init__()
+        assert covariance_type in ("full",)
+        self.covariance_type = covariance_type
+        self.n_components = n_components
+        self.n_features = n_features
+        self.seed = seed
+
+        self.register_buffer("means", torch.zeros(n_components, n_features))
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:86]: can set these to `torch.Tensor()` (the empty tensor) by default instead, then can check if fitted in froward

File: nequip/utils/gmm.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        self.register_buffer(
+            "precisions_cholesky", torch.zeros(n_components, n_features, n_features)
+        )
+        # TODO: other parameters as buffers
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Compute the NLL of samples ``x`` under the GMM."""
+        # TODO: implement
+
+        # TODO: check if model has been fitted
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:100]: check if buffer is empty (`x.numel == 0`)

File: nequip/utils/gmm.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+        # TODO: testing
+        estimated_log_probs = _estimate_log_gaussian_prob(
+            x, self.means, self.precisions_cholesky, self.covariance_type
+        )
+
+        estimated_weights = np.log(self.weights)
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:107]:
```suggestion
        estimated_weights = torch.log(self.weights)
```

File: nequip/utils/gmm.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    def fit(self, x: torch.Tensor) -> None:
+        """Fit the GMM to the samples `x` using sklearn."""
+        # TODO: if n_components is None, use the BIC; else just use provided n_components
+        #       in the BIC case, make sure to set it to an int of the final n_components
+        # TODO: fit with sklearn and set this objects buffers from sklearn values
+        # Set number of n_components if given, otherwise use the BIC
+
+        n_components = self.n_components
+        if not n_components:
+            tmp_heurstic = 30  # TODO: determine how to set max number of components for BIC testing
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:121]: add a parameter to `fit` with some sane-ish default of like 30 or 50?

File: nequip/utils/gmm.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+            components = np.arange(1, tmp_heurstic)
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:122]: can just use `range` here, better form since doesn't allocate an array
(not actually a problem, just a code style thing)

File: nequip/utils/gmm.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            gmms = [
+                mixture.GaussianMixture(
+                    n_components=n, covariance_type="full", random_state=27617
+                )
+                for n in components
+            ]
+            bics = [model.fit(x).bic(x) for model in gmms]
+            n_components = bics.index(min(bics))
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:130]:
```suggestion
            n_components = np.argmin(bics)
```

File: nequip/utils/gmm.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        # TODO: fit with sklearn and set this objects buffers from sklearn values
+        # Set number of n_components if given, otherwise use the BIC
+
+        n_components = self.n_components
+        if not n_components:
+            tmp_heurstic = 30  # TODO: determine how to set max number of components for BIC testing
+            components = np.arange(1, tmp_heurstic)
+            gmms = [
+                mixture.GaussianMixture(
+                    n_components=n, covariance_type="full", random_state=27617
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:125]:
```suggestion
                    n_components=n, covariance_type="full", random_state=torch.randint(2**16, (1,)).item()
```
this way the seed comes from torch, and is indirectly seeded by the pytorch seed? idk

File: nequip/utils/gmm.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                )
+                for n in components
+            ]
+            bics = [model.fit(x).bic(x) for model in gmms]
+            n_components = bics.index(min(bics))
+
+        # Fit GMM
+        random_state = self.seed if self.seed else 123
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:133]: see above, I would remove `self.seed` and just allow a `torch.Generator` to be passed to this function optionally. (If nothing, just call `torch.Generator()`.)

File: nequip/utils/gmm.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+        # Save info from GMM into the register buffers
+        self.register_buffer("means", torch.from_numpy(gmm.means_))
+        self.register_buffer("weights", torch.from_numpy(gmm.weights_))
+        self.register_buffer("covariances", torch.from_numpy(gmm.covariances_))
+        self.register_buffer(
+            "precisions_cholesky", torch.from_numpy(gmm.precisions_cholesky_)
+        )
+
+        # TODO: replace above code with self.means[:] = gmm.means_ ?
```
Comment (Linux-cpp-lisp) [inline nequip/utils/gmm.py:149]: I think this is fine, in particular if we don't initialize to zeros as above comment

File: tests/unit/utils/test_gmm.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+import numpy as np
+from nequip.utils import gmm
+from sklearn import mixture
+
+fit_data = np.random.rand(10, 3)
+gmm_sklearn = mixture.GaussianMixture(
+    n_components=3, covariance_type="full", random_state=123
+)
+
+gmm_torch = gmm.GaussianMixture(n_components=3, seed=123, n_features=3)
```
Comment (Linux-cpp-lisp) [inline tests/unit/utils/test_gmm.py:11]: Use pytest fixtures to construct common data or objects used in multiple tests:  https://docs.pytest.org/en/7.1.x/explanation/fixtures.html#about-fixtures

File: tests/unit/utils/test_gmm.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+# gmm_torch = torch.jit.script(gmm_torch)
```
Comment (Linux-cpp-lisp) [inline tests/unit/utils/test_gmm.py:12]: You can make a test that uses `assert_auto_jitable` to check whether your thing can be compiled correctly: https://github.com/mir-group/nequip/blob/fe73530ea34b16143de60cd9ed5e1dd14ce625b4/tests/unit/nn/test_embed.py#L19

File: tests/unit/utils/test_gmm.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class TestGMM:
+    def test_fit_forward_simple(self):
+        gmm_sklearn.fit(fit_data)
+        gmm_torch.fit(torch.from_numpy(fit_data))
+
+        assert torch.allclose(
+            torch.from_numpy(gmm_sklearn.means_), gmm_torch.state_dict()["means"]
```
Comment (Linux-cpp-lisp) [inline tests/unit/utils/test_gmm.py:21]:
```suggestion
            torch.from_numpy(gmm_sklearn.means_), gmm_torch.means
```
no need to check through state dict

File: tests/unit/utils/test_gmm.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        assert torch.allclose(
+            torch.from_numpy(gmm_sklearn.weights_), gmm_torch.state_dict()["weights"]
+        )
+
+        assert torch.allclose(
+            torch.from_numpy(gmm_sklearn.precisions_cholesky_),
+            gmm_torch.state_dict()["precisions_cholesky"],
+        )
+
+        test_data = np.random.rand(20, 3)
```
Comment (Linux-cpp-lisp) [inline tests/unit/utils/test_gmm.py:38]: Always use some fixed seed you make up for unit tests; otherwise it is extremely difficult to reproduce failures which can be very confusing. (Can also lead to nondeterminism with regard to test ordering, which is bad; for example on GitHub actions we run our tests in two threads, so they run in a different order than when you run them in one thread locally)

File: tests/unit/utils/test_gmm.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+        sklearn_nll = gmm_sklearn.score_samples(test_data)
+        torch_nll = gmm_torch.forward(torch.from_numpy(test_data))
```
Comment (Linux-cpp-lisp) [inline tests/unit/utils/test_gmm.py:41]:
```suggestion
        torch_nll = gmm_torch(torch.from_numpy(test_data))
```
`forward` is called automatically by calling a `torch.nn.Module`

File: nequip/utils/gmm.py [hunk 18]
```diff
@@ -55,7 +69,7 @@ def _estimate_log_gaussian_prob(
 
     # Since we are using the precision of the Cholesky decomposition,
     # `-0.5 * log_det` becomes `+ log_det`
-    return -0.5 * (n_features * torch.log(torch.tensor(2 * np.pi)) + log_prob) + log_det
+    return -0.5 * (n_features * math.log(math.pi) + log_prob) + log_det
```
[the remark this answers is not in the document]
Comment (albertzhu01) [reply nequip/utils/gmm.py]: Yep my bad, added it back
